pymaker/dss.py

```diff
@@ -827,49 +827,80 @@ def __init__(self, web3: Web3, address: Address):
     def live(self) -> bool:
         return self._contract.functions.live().call() > 0
 
+    def can_bite(self, ilk: Ilk, urn: Urn) -> bool:
+        """ Determine whether a vault can be liquidated
+
+        Args:
+            ilk: Collateral type
+            urn: Identifies the vault holder or proxy
+        """
+        assert isinstance(ilk, Ilk)
+        assert isinstance(urn, Urn)
+        ilk = self.vat.ilk(ilk.name)
+        urn = self.vat.urn(ilk, urn.address)
+        rate = self.vat.ilk(ilk.name).rate
+
+        # Collateral value should be less than the product of our stablecoin debt and the debt multiplier
+        safe = Ray(urn.ink) * ilk.spot >= Ray(urn.art) * rate
+        if safe:
+            return False
+
+        # Ensure there's room in the litter box
+        box: Rad = self.box()
+        litter: Rad = self.litter()
+        room: Rad = box - litter
+        if litter >= box or room < ilk.dust:
+            return False
+
+        # Prevent null auction (ilk.dunk [Rad], ilk.rate [Ray], ilk.chop [Wad])
+        dart: Wad = min(urn.art, Wad(min(self.dunk(ilk), room) / Rad(ilk.rate) / Rad(self.chop(ilk))))
+        dink: Wad = min(urn.ink, urn.ink * dart / urn.art)
+
+        return dart > Wad(0) and dink > Wad(0)
+
     def bite(self, ilk: Ilk, urn: Urn) -> Transact:
-        """ Initiate liquidation of a CDP, kicking off a flip auction
+        """ Initiate liquidation of a vault, kicking off a flip auction
 
         Args:
             ilk: Identifies the type of collateral.
-            urn: Address of the CDP holder.
+            urn: Address of the vault holder.
         """
         assert isinstance(ilk, Ilk)
         assert isinstance(urn, Urn)
 
         ilk = self.vat.ilk(ilk.name)
         urn = self.vat.urn(ilk, urn.address)
         rate = self.vat.ilk(ilk.name).rate
-        logger.info(f'Biting {ilk.name} CDP {urn.address.address} with ink={urn.ink} spot={ilk.spot} '
+        logger.info(f'Biting {ilk.name} vault {urn.address.address} with ink={urn.ink} spot={ilk.spot} '
                     f'art={urn.art} rate={rate}')
```

Comment:
My working copy of the keeper currently calls `can_bite()` for every urn to find opportunities to bite.  We wouldn't want to duplicate the 2-6 onchain calls.  I named it `can_bite` rather than `validate_bite` in an attempt to convey this.

Reply:
Got it. The onus will be on the user to check `can_bite`. 👍